### PR TITLE
new package: 2D Movable Separability of Sets (2D Casting)

### PR DESCRIPTION
**Pull request by efifogel**
## Summary of Changes

This is a new package called "2D Movable Separability of Sets".

## Release Management

* Affected package(s): Set_movable_separability_2
* Issue(s) solved (if any): fix #0000, fix #0000,...
* Feature: [Features/Casting 2][1]

  [1]: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Casting_2

### CODE REVIEW

**Comment by lrineau** [comment]
The feature is named "Casting 2" but the new package directory is `Casting_3`. Is that an error?

**Comment by lrineau** [comment]
TODO for later, before the integration: update the patch to `changes.html` so that the new section is added to the changelog for CGAL-4.12 (instead of 4.11).

**Comment by sloriot** [comment]
@efifogel the travis job failed.

**Comment by sloriot** [comment]
@efifogel There is a `Casting_3` directory added in this PR. I think it is a mistake. There is also a Casting_3 in the title of the PR. It is also a mistake I guess. Travis is reporting some compilation errors in the arrangement package, Envelope_3, ...

**Comment by efifogel** [comment]
- I removed the Casting_3 directory.
- Indeed, the 'Casting_3' in the title is a mistake. It should say
'Casting_2'.
- How can I see the Travis report?

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
Good! I edit the PR name. To get access to the travis report click on _Details_, then all red lines indicates that they were an error in a build. SInce you pushed new commits there are none for now but you can get access to the previous [one](https://travis-ci.org/CGAL/cgal/builds/262550673)

**Comment by sloriot** [comment]
There are a few errors and warnings. See [test results](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-17.shtml)

**Comment by efifogel** [comment]
Fixed, committed, and pushed.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
There are still some errors. See [test results](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-18.shtml)

**Comment by lrineau** [comment]
Travis reports an error:
```
The matrix and the actual package list differ : .
Set_movable_separability_2
```
https://travis-ci.org/CGAL/cgal/jobs/265871519#L6781

@maxGimeno, can you explain what has to be done?

**Comment by efifogel** [comment]
It's not something that I should fix, right?

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
@efifogel: The list of CGAL packages is hard-coded somewhere in our Travis infrastructure, for an extra verification. Maxime (@maxGimeno) will explain what has to be done to add a new package, and one of us will commit the change.

**Comment by maxGimeno** [comment]
I think you just need to call .travis/generate_travis.sh, so the package list is updated in the .travis.yml.
 I will add a section in the guidelines.

**Comment by lrineau** [comment]
This PR was in `integration` last might, and the documentation testsuite reported one error:
```
.../CGAL-4.12-Ic-27/doc/Set_movable_separability_2/Set_movable_separability_2.txt:84: Warning: unable to resolve reference to `Chapter_2D_Regularized_Boolean_Set-Operations' for \ref command
```
https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.12-Ic-27/Set_movable_separability_2.log

**Comment by lrineau** [comment]
... and the code testsuite showed multiple errors, on Windows, and with old Linux compiler:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-27.shtml#Set_movable_separability_2

**Comment by afabri** [comment]
Concerning the link to the other package it is in general a better option to `\ref` to the `Pkg...Summary` anchor in the `PackageDescription.txt` file.

**Comment by afabri** [comment]
is the file name too long???

**Comment by efifogel** [comment]
It turns out that a reference to a chapter is resolved only if the package
is listed in the "dependencies" file.
I've added the package 'Boolean_set_operations_2' to the dependencies file,
and indeed it fixed the 'unresolved' error.
The only problem left is that there is no dependency code wise. That is,
the code of Set_movable_separability_2 does not depend on the code of
Boolean_set_operations_2. I hope this does not cause a problem when a
customer is interested in purchasing a license for
Set_movable_separability_2.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
The compilation errors on Windows are probably due to the long file name.

The compilation errors on Linux are due to the requirement of C++11, that is not correctly handled: it is currently mandatory, and there is nothing that disables the compilation of this directory when the compiler cannot enable C++11.

**Comment by efifogel** [comment]
On 29 August 2017 at 18:37, Laurent Rineau <notifications@github.com> wrote:

> The compilation errors on Windows are probably due to the long file name.
>
This is a bummer. Anyhow, I've shorten the file names.

> The compilation errors on Linux are due to the requirement of C++11, that
> is not correctly handled: it is currently mandatory, and there is nothing
> that disables the compilation of this directory when the compiler cannot
> enable C++11.
>

The conditional statement that guards for C++11 feature is there, but it
wasn't respected. I've bumped up the required cmake version. I hope that
this fixes it.

All pushed.

File: Set_movable_separability_2/test/Set_movable_separability_2/CMakeLists.txt [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 project( Set_movable_separability_2_Tests )
 
-cmake_minimum_required(VERSION 2.8.10)
+cmake_minimum_required(VERSION 3.1)
```
Comment (lrineau) [inline Set_movable_separability_2/test/Set_movable_separability_2/CMakeLists.txt]: What is missing is the following:
```cmake
list(FIND CMAKE_CXX_COMPILE_FEATURES cxx_generalized_initializers has_cpp11)
if(has_cpp11 LESS 0)		  if(has_cpp11 LESS 0)
   message(STATUS "NOTICE: These examples requires a C++11 compiler and will not be compiled.")		      
   return()
endif()
```

**Comment by efifogel** [comment]
You are right.
It's fixed now.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
Doxygen warnings:
```
/home/cgal-testsuite/cgal_doc_build/CGAL-4.12-Ic-29/doc/Set_movable_separability_2/examples.txt:6: Warning: included file Set_movable_separability_2/is_pullout_direction_single_mold_translational_casting.cpp is not found. Check your EXAMPLE_PATH
/home/cgal-testsuite/cgal_doc_build/CGAL-4.12-Ic-29/doc/Set_movable_separability_2/examples.txt:6: Warning: included file Set_movable_separability_2/pullout_directions_single_mold_translational_casting.cpp is not found. Check your EXAMPLE_PATH
/home/cgal-testsuite/cgal_doc_build/CGAL-4.12-Ic-29/doc/Set_movable_separability_2/examples.txt:7: Warning: included file Set_movable_separability_2/top_edges_single_mold_translational_casting.cpp is not found. Check your EXAMPLE_PATH
/home/cgal-testsuite/cgal_doc_build/CGAL-4.12-Ic-29/doc/Set_movable_separability_2/Set_movable_separability_2.txt:103: Warning: included file Set_movable_separability_2/top_edges_single_mold_translational_casting.cpp is not found. Check your EXAMPLE_PATH
```
https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.12-Ic-29/Set_movable_separability_2.log

**Comment by lrineau** [comment]
And there are warnings in the code testsuite:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-29.shtml#Set_movable_separability_2

**Comment by efifogel** [comment]
Fixed

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by efifogel** [comment]
This is fixed as well.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
Efi, you still have to run `.travis/generate_travis.sh` from the root of the Git repository, and commit and push the change to `.travis.yml`. That will fix the [Travis `CHECK` line][1], and make Travis actually test the new packages along with the others.

[1]: https://travis-ci.org/CGAL/cgal/jobs/270448050#L6764

File: Set_movable_separability_2/test/Set_movable_separability_2/CMakeLists.txt [hunk 2]
```diff
@@ -18,19 +17,20 @@ set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
...
 else()
-
-    message(STATUS "This program requires the CGAL library, and will not be compiled.")
-
+  if (RUNNING_CGAL_AUTO_TEST)
+    # Just to suppress a warning from CMake when that variable is set on the command line...
+  endif()
+  if (CGAL_DIR)
+    # Just to suppress a warning from CMake when that variable is set on the command line...
+  endif()
```
Comment (lrineau) [inline Set_movable_separability_2/test/Set_movable_separability_2/CMakeLists.txt]: [The CMake warnings are still threre][1]. The dummy uses of `RUNNING_CGAL_AUTO_TEST` and `CGAL_DIR` should be before the `return()` [line 11][l11].

[1]: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.12-Ic-30/Set_movable_separability_2/TestReport_lrineau_CentOS6-32.gz
[l11]: https://github.com/CGAL/cgal/pull/2259/commits/99a673fd0e2d4b345037bdd070fb2b049fe3c71e#diff-89ea674154596b8f9e8a91734fee530dR11

**Comment by efifogel** [comment]
When I just run .travis/generate_travis.sh the file .travis.yml does not
change, but I get the following messages:

cat: /home/efif/trees/.travis/template.txt: No such file or directory
cat: /home/efif/trees/.travis/template.txt: No such file or directory

Should I add Set_movable_separability_2 to .travis/packages.txt?

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
Actually the script has to be run from inside the `.travis/` subdirectory. I did so in the commit 
https://github.com/CGAL/cgal/pull/2259/commits/c84964d0ae7f13ba4c21d78ed9548d7f2510dc22. The branch should be ready, now.

**Comment by lrineau** [comment]
This package/pull-request is now fine in both Travis and [the testsuite][1]. Congrats.

I will merge it soon in `master`.

[1]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-31.shtml#Set_movable_separability_2

**Comment by lrineau** [comment]
@sloriot: the size of this PR is rather heavy:
```
renoir ~/Git/cgal $ ./Scripts/developer_scripts/test_merge_of_branch https://github.com/CGAL/cgal/pull/2259
[...]
.. Size of the gzipped bundle
=> 1306364
```

But if the history is squashed, the size is only 74530. Can you try to modify the history so that the full history has a lower weight?

**Comment by sloriot** [comment]
There were 2 executables. Removing them reduces the size to 224Ko.

@efifogel Before commiting anything to this branch, you need to reset your local branch.

**Pull request merged**